onnx_coreml/_operators.py

```diff
@@ -414,28 +414,29 @@ def _convert_concat(builder, node, graph, err):  # type: (NeuralNetworkBuilder,
 
 def _convert_reduce(builder, node, graph, err):  # type: (NeuralNetworkBuilder, Node, Graph, ErrorHandling) -> None
 
-    axes = node.attrs['axes']
+    axes = node.attrs.get('axes', None)
+    if axes is None:
+        assert node.inputs[0] in graph.shape_dict, "Shape inference failed for reduce op"
+        shape = graph.shape_dict[node.inputs[0]]
+        axes = range(0, len(shape))
 
     def get_coreml_axis(axes): # type: (List[int]) -> Text
         coreml_axis = ""
-        if node.inputs[0] in graph.shape_dict:
-            input_shape = graph.shape_dict[node.inputs[0]]
-            if len(input_shape) == 1: coreml_axis = 'C'
-            elif len(input_shape) == 2:
-                if len(axes) == 1 and axes[0] == 1: coreml_axis = 'C'
-            elif len(input_shape) == 3:
-                for ind in [['C','H','W'][i] for i in axes]: coreml_axis += ind
-            elif len(input_shape) == 4:
-                for ind in [['B','C','H','W'][i] for i in axes]: coreml_axis += ind
-        else:  # shape info is not available. Fall back to guessing (ideally this should not happen)
+        assert node.inputs[0] in graph.shape_dict, "Shape inference failed for reduce op"
```

Comment:
By using this assert, we are banking on the fact that shape info will always be available?

Reply:
Yes. I had to add some code above (the one handling axes) that also depends on shape info. I think it's better to error out than to assume some shape which can be false and the user gets wrong behavior without knowing it. What do you think?

Reply:
I think we should keep the default logic in, as removing it might break some models (maybe those that we are not testing in travis).
Once the shape info feature is properly in onnx, that code path will anyways be not used and can be removed later.